### PR TITLE
Extend documentation content and fix code examples

### PR DESCRIPTION
**Pull request by Ph0enixKM**
- Improved documentation on basic syntax also explaining better the limitations of Bash
- Visual changes and improvements
- New details card
- Added constants section
- Made documentation sound more professional

### CODE REVIEW

**Review by Mte90 — APPROVED** [review]

**Comment by Ph0enixKM** [comment]
Here is a preview of the newely added `> DETAILS:` block

Screenshot 2024-12-19 at 16 48 35

**Comment by Ph0enixKM** [comment]
@hdwalters could you review the markdown files? I read through them but I think it would be nice if you could take a quick look as well

File: docs/0.3.5-alpha/advanced_syntax/as_cast.md [hunk 1]
```diff
@@ -2,7 +2,7 @@ As cast is a tool that might initially seem like an easy way to convert types. S
 
 ## Regular Casts
 
-There might be times when you want to pass a variables that is a `Bool` to a function that accepts `Num`. Since Bool and Num are types that are compatible with each other, you can easily cast one into the other like so:
+There might be times when we want to pass a variables that is a `Bool` to a function that accepts `Num`. Since Bool and Num are types that are compatible with each other, you can easily cast one into the other like so:
```
Comment (hdwalters) [inline docs/0.3.5-alpha/advanced_syntax/as_cast.md:5]: Typo: "pass a variable".

File: docs/0.3.5-alpha/by_example/ubuntu_updater.md [hunk 2]
```diff
@@ -16,7 +14,7 @@ main {
   silent unsafe $ iwgetid -r | grep -E '(OEBB|WESTlan)' $
   if status == 0 {
     echo "Skipping updates because of slow Wifi"
-    exit(0)
+    exit 0
```
Comment (hdwalters) [inline docs/0.3.5-alpha/by_example/ubuntu_updater.md:17]: Wasn't `exit` still in the standard library in 0.3.5?

File: docs/0.4.0-alpha/basic_syntax/arrays.md [hunk 3]
```diff
@@ -53,3 +53,44 @@ let cities = capitals + ["Barcelona", "Florence"]
 ```
 
 In order to see more operations on the array data type take a look at the standard library documentation which covers functions such as `join`, `len` or `sum`.
+
+## Nested arrays
```
Comment (hdwalters) [inline docs/0.4.0-alpha/basic_syntax/arrays.md:57]: Suggest capitalising "Arrays".

File: docs/0.4.0-alpha/basic_syntax/arrays.md [hunk 4]
```diff
@@ -53,3 +53,44 @@ let cities = capitals + ["Barcelona", "Florence"]
...
+typedef struct array_element {
+	arrayind_t	ind;
+	char	*value;
+#ifndef ALT_ARRAY_IMPLEMENTATION
+	struct array_element *next, *prev;
+#endif
+} ARRAY_ELEMENT;
+```
+
+# Ranges
```
Comment (hdwalters) [inline docs/0.4.0-alpha/basic_syntax/arrays.md:84]: I moved this section to the top, in order to reference it in the array slicing section. Need to remove this duplicate.

File: docs/0.4.0-alpha/basic_syntax/basic_syntax.md [hunk 5]
```diff
@@ -4,8 +4,11 @@ Since Amber was designed based on the ECMA script syntax some of the following p
 
 > We suggest to take a look on [the examples](https://docs.amber-lang.com/by_example/examples) we provide to see real use cases written in Amber.
 
+Here you might notice an `echo` built-in that does the same thing that Bash's `echo` do.
+
 Here is a code snippet that illustrates certain features of Amber. We will provide detailed explanations for each of these features and cover additional content in the forthcoming chapters.
 
+
```
Comment (hdwalters) [inline docs/0.4.0-alpha/basic_syntax/basic_syntax.md:11]: Unnecessary blank line.

File: docs/0.4.0-alpha/basic_syntax/conditions.md [hunk 6]
```diff
@@ -5,7 +5,7 @@ There are three ways to perform conditional logic:
 
 ## If Statement
 
-The good ol' if statement that you may recognize from other modern programming languages:
+The good ol' if statement that one may recognize from other modern programming languages:
```
Comment (hdwalters) [inline docs/0.4.0-alpha/basic_syntax/conditions.md:8]: Perhaps "good ol'" is a bit folksy?

File: docs/0.4.0-alpha/basic_syntax/conditions.md [hunk 7]
```diff
@@ -37,6 +37,7 @@ else:
 	echo "Entry not allowed"
 ```
 
+
```
Comment (hdwalters) [inline docs/0.4.0-alpha/basic_syntax/conditions.md:40]: Unnecessary blank line.

File: docs/0.4.0-alpha/basic_syntax/data_types.md [hunk 8]
```diff
@@ -1,4 +1,6 @@
-Amber supports 5 data types:
+In Bash there is only one primitive data type, string, which internal implementation is represented by an array of characters `char*`. Amber extends on this data type to introduce a couple more.
```
Comment (hdwalters) [inline docs/0.4.0-alpha/basic_syntax/data_types.md:1]: Suggest "to introduce a few more"; "couple" implies two!

File: docs/0.4.0-alpha/basic_syntax/data_types.md [hunk 9]
```diff
@@ -9,11 +11,15 @@ Amber supports 5 data types:
...
 `Text` data type is the most basic data type in Amber. It's just a string of characters and is also stored as a string of characters under the hood.
 
+Text literal in Amber is contained between double quotes. Amber makes sure to prevent content inside from [globbing](https://en.wikipedia.org/wiki/Glob_%28programming%29). This prevents unexpected behaviors from happening.
+
 ```ab
 // `Text` literal:
 "Welcome to the jungle"
 ```
 
+Just like in other programming languages, characters in Text literals can be escaped.
```
Comment (hdwalters) [inline docs/0.4.0-alpha/basic_syntax/data_types.md:21]: Should probably put "Text" in backticks.

File: docs/0.4.0-alpha/basic_syntax/data_types.md [hunk 10]
```diff
@@ -28,14 +34,14 @@ Under the hood its value is stored as a string of characters - the same way as i
 
 ## Boolean
 
+Boolean values are translated to `0` or `1` numerical values. These values can be easily [casted](/advanced_syntax/as_cast) to numbers `Num`.
```
Comment (hdwalters) [inline docs/0.4.0-alpha/basic_syntax/data_types.md:37]: I would put "can easily be cast".

File: docs/0.4.0-alpha/basic_syntax/data_types.md [hunk 11]
```diff
@@ -47,14 +53,20 @@ The most common use of this data type is when you want to declare that some func
...
-When dealing with arrays you probably want to define of which data type you want to create the array. In that case simply put the data type inside of the square brackets. For example an array of type `Num` is `[Num]` (in C like languages it would be `Num[]`)
+Arrays in Amber and Bash are dynamically allocated. When creating an array literal it is important to specify of which data type the array should be made. Type signature of arrays can be represented with `[T]` where `T` is the value type that this array holds. Example: an array of type `Num` is `[Num]` (in C like languages it would be `Num[]`).
+
+To create an array literal simply enclose a list of elements separated with a comma `,` with square brackets `[]`.
 
 ```ab
 // `[Num]` data type
 [1, 2, 3]
+
+
```
Comment (hdwalters) [inline docs/0.4.0-alpha/basic_syntax/data_types.md:64]: Unnecessary blank line.

File: docs/0.4.0-alpha/basic_syntax/expressions.md [hunk 12]
```diff
@@ -73,7 +73,23 @@ echo age // Outputs: 23
 
 ## Text Interpolation
 
-There are moments when you want to embed values into a text:
+Text interpolation is a form of embedding various values into the text literal that are combined together with their textual representations.
```
Comment (hdwalters) [inline docs/0.4.0-alpha/basic_syntax/expressions.md:76]: I would put "that are combined with".

File: docs/0.4.0-alpha/basic_syntax/importing.md [hunk 13]
```diff
@@ -31,24 +31,24 @@ echo sum(1, sub(2, mul(4, 5)))
...
+This statement imports all functions defined in file.ab and re-exports them, making them publicly accessible from the current file.
 
 ## Main Block
 
-Sometimes you want to run certain code when file is being run directly. The same issue can be solved in Python:
+In case when we want a specific code to run only when a file is executed directly, Amber offers a clean and powerful solution. Similar to Python’s approach:
 
 ```py
 if __name__ == '__main__':
 	# code to execute
```
Comment (hdwalters) [inline docs/0.4.0-alpha/basic_syntax/importing.md:48]: Suggest replacing the tab with four spaces.

File: docs/0.4.0-alpha/basic_syntax/importing.md [hunk 14]
```diff
@@ -59,6 +59,11 @@ main {
 }
 ```
 
+> DETAILS: Key features of `main` block:
+- Code outside the main block runs regardless of how the file is executed.
+- Code inside the main block runs only when the file is executed directly.
+- The `?` operator ensures that any exit code from some command is automatically passed back to the shell, making it easy to handle script results effectively.
```
Comment (hdwalters) [inline docs/0.4.0-alpha/basic_syntax/importing.md:65]: Do we state anywhere that this also applies to defined functions?

File: docs/0.4.0-alpha/basic_syntax/importing.md [hunk 15]
```diff
@@ -79,7 +84,7 @@ Main block can provide an array of arguments (that is of type `[Text]`) passed t
 
 ```ab
 main (args) {
-	loop i, arg in args {
+	for i, arg in args {
```
Comment (hdwalters) [inline docs/0.4.0-alpha/basic_syntax/importing.md:87]: Should probably replace tabs with spaces.

File: docs/0.4.0-alpha/basic_syntax/variables.md [hunk 16]
```diff
@@ -27,6 +29,11 @@ let result = 123
...
 ```
 
-## Reserved Prefix
+# Constant
+
+Constant is a type of variable that cannot be modified.
 
-The Amber compiler reserves all identifiers starting with double underscore `__` in addition to keywords like `let`, `if`, etc.
+```ab
+const gravity = 9.81
```
Comment (hdwalters) [inline docs/0.4.0-alpha/basic_syntax/variables.md:37]: I love this example!  But if you really wanted a universal constant, you could use `speed_of_light`.

File: docs/0.4.0-alpha/by_example/lsp_installer.md [hunk 17]
```diff
@@ -4,48 +4,48 @@ This script automates the installation of several Language Server Protocol (LSP)
...
+            if text_contains("tar.gz", packed_file) {
                 $ tar -zxvf "./{packed_file}" -C ./ > /dev/null 2>&1 $
                 trust mv "./{binary}" "/usr/local/bin"
             } else {
                 $ gunzip -c - > "/usr/local/bin/{binary}" $
             }
             $ rm "./{packed_file}" $
         }
-        make_executable("/usr/local/bin/{binary}")
+        chmod("/usr/local/bin/{binary}", "+x")
```
Comment (hdwalters) [inline docs/0.4.0-alpha/by_example/lsp_installer.md:45]: Should be `file_chmod`.

File: src/components/Markdown/amber.ts [hunk 18]
```diff
@@ -31,7 +31,7 @@ const amber = (hljs: HLJSApi) => {
     }
     const keywords = {
         scope: 'keyword',
-        match: /\b(and|as|break|cd|continue|echo|else|fail|failed|for|from|fun|if|import|in|is|len|let|lines|loop|main|mv|nameof|not|or|pub|ref|return|silent|status|then|trust|unsafe)\b/
+        match: /\b(ref|to|error|status|if|echo|loop|for|in|silent|return|fun|else|break|continue|and|or|not|let|const|sh|main|failed|fail|unsafe|then|pub|import|from|as|nameof|is)\b/
```
Comment (hdwalters) [inline src/components/Markdown/amber.ts:34]: Should probably put this back into alphabetical order, for simpler comparison next time round.

**Review by hdwalters — CHANGES_REQUESTED** [review]
I'll push a new commit with suggested changes, to keep things moving. Hope this is okay!

File: docs/0.3.5-alpha/by_example/ubuntu_updater.md (hunk 2 above)
Comment (Mte90) [reply docs/0.3.5-alpha/by_example/ubuntu_updater.md:17]: I don't think that was in 0.3.5

Comment (Ph0enixKM) [reply docs/0.3.5-alpha/by_example/ubuntu_updater.md:17]: Ah you're right @hdwalters @Mte90 I thought that it was already a part of 0.3.5. Double checked on that and it seems that it was not. I'll fix this

File: docs/0.4.0-alpha/basic_syntax/importing.md (hunk 14 above)
Comment (Ph0enixKM) [reply docs/0.4.0-alpha/basic_syntax/importing.md:65]: Could you elaborate?

File: docs/0.4.0-alpha/basic_syntax/variables.md (hunk 16 above)
Comment (Ph0enixKM) [reply docs/0.4.0-alpha/basic_syntax/variables.md:37]: Good idea!

File: docs/0.3.5-alpha/basic_syntax/data_types.md [hunk 19]
```diff
@@ -9,11 +11,15 @@ Amber supports 5 data types:
...
 `Text` data type is the most basic data type in Amber. It's just a string of characters and is also stored as a string of characters under the hood.
 
+Text literal in Amber is contained between double quotes. Amber makes sure to prevent content inside from [globbing](https://en.wikipedia.org/wiki/Glob_%28programming%29). This prevents unexpected behaviors from happening.
+
 ```ab
 // `Text` literal:
 "Welcome to the jungle"
 ```
 
+Just like in other programming languages, characters in Text literals can be escaped.
```
Comment (Ph0enixKM) [inline docs/0.3.5-alpha/basic_syntax/data_types.md:21]:
```suggestion
Just like in other programming languages, characters in `Text` literals can be escaped.
```

**Review by hdwalters — APPROVED** [review]

**Pull request merged**